flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -485,6 +494,23 @@ public void requestPartitions() throws IOException, InterruptedException {
 		}
 	}
 
+	@Override
+	public void blockInputChannel(int channelIndex) {
+		InputChannel inputChannel = indexToInputChannelMap.get(channelIndex);
+		if (inputChannel == null) {
+			throw new IllegalStateException("Could not find input channel from the channel index " + channelIndex);
+		}
+
+		inputChannel.setBlocked(true);
+	}
+
+	@Override
+	public void releaseBlockedInputChannels() {
+		for (InputChannel inputChannel : inputChannels.values()) {
+			inputChannel.setBlocked(false);
+		}
```

Comment:
~Do we need to make sure that there's no concurrent `blockInputChannel` call trying to block a channel for an alignment for a later checkpoint or can we assume here that we are still processing one alignment (doing the release) and therefore cannot concurrently block?~
I guess we are safe here - `BarrierBuffer` actually does it the same way.